apis/metal3.io/v1alpha1/baremetalhost_types.go

```diff
@@ -869,6 +884,11 @@ func (host *BareMetalHost) NeedsProvisioning() bool {
 		// The host is not supposed to be powered on.
 		return false
 	}
+
+	return host.hasNewImage() || host.hasNewCustomDeploy()
```

Comment:
I'm not sure we really want the new spec field to trigger reprovisioning?

Currently the only field which does that is the image, so as an existing user I might be surprised if I changed some value in the custom deploy field and all my hosts got rebuilt?

I'm thinking about consistency e.g with userData/networkData etc where those fields (or rather the secrets they reference) could change post-deploy, but we don't rebuild even though they influence the deployed OS - this is similar even though it won't reference a potentially-shared secret?

Reply:
If we don't do this, how will we unprovision? In case when Image == nil and CustomDeploy != nil, the only way to unprovision is to set CustomDeploy = nil.

Reply:
Perhaps I'm confused, can you outline the expected provision/deprovision flow here?

I expect a BMH to deprovision only when I remove (or replace) the image URL, so from a users perspective I'd expect the same behavior even if I have chosen to use a custom deploy ramdisk?

Reply:
Yes, but you don't have an image now. What would you do to deprovision?

My assumption was that a user removes CustomDeploy from Spec, similarly to how Image removal causes deprovisioning now.

Reply:
In the current flow, if the `spec.image` gets removed, that triggers deprovisioning - in the case where CustomDeploy is specified I'd expect the same behavior.

But I probably would *not* expect to trigger rebuild of the host, if I modify CustomDeploy on an already provisioned host - most other spec fields don't trigger rebuild, so I think I'd expect this to behave the same.

If someone removes CustomDeploy from the spec, I'd expect to revert to the existing behavior (use the default deploy steps and IPA kernel/ramdisk) - obviously there are likely cases where it doesn't make sense to remove CustomDeploy because that configuration is required for provisioning and deprovisioning, but that's up to the user?

Reply:
Sorry, I don't get it. How do you suggest we trigger reprovision if Image is nil already?

Reply:
> use the default deploy steps and IPA kernel/ramdisk

We have untied this logic from custom image logic, so only the default steps. But then we end up with a provisioned node with Image==nil and CustomDeploy==nil?

Reply:
Ok we discussed this, and the reason for my confusion is that it's now possible to have `CustomDeploy!=nil`, but with `Image==nil` e.g the Image field is now optional (which is mentioned in the [design](https://github.com/metal3-io/metal3-docs/blob/master/design/deploy-steps.md))

That is unfortunate since it changes the existing API behavior, and potentially means more code corner cases since we have two APIs now that trigger provisioning/deprovisioning.

I know @zaneb had some suggestions on using the image URL instead, but since we accepted the design with this API I'm not blocking on it unless others think we need to revisit that discussion.